app/src/main/java/com/kickstarter/viewmodels/SignupViewModel.java

```diff
@@ -13,153 +13,188 @@
-  }
+    /** Call when the password field changes. */
+    void password(String password);
 
-  // ERRORS
-  private final PublishSubject<ErrorEnvelope> signupError = PublishSubject.create();
-  public Observable<String> signupError() {
-    return this.signupError
-      .takeUntil(this.signupSuccess)
-      .map(ErrorEnvelope::errorMessage);
-  }
+    /** Call when the send newsletter toggle changes. */
+    void sendNewslettersClick(boolean send);
 
-  public final SignupViewModelInputs inputs = this;
-  public final SignupViewModelOutputs outputs = this;
-  public final SignupViewModelErrors errors = this;
-
-  public SignupViewModel(final @NonNull Environment environment) {
-    super(environment);
-
-    this.client = environment.apiClient();
-    this.currentConfig = environment.currentConfig();
-    this.currentUser = environment.currentUser();
-
-    final Observable<SignupData> signupData = Observable.combineLatest(
-      this.fullName, this.email, this.password, this.sendNewslettersIsChecked, SignupData::new
-    );
-
-    this.sendNewslettersClick
-      .compose(bindToLifecycle())
-      .subscribe(this.sendNewslettersIsChecked::onNext);
-
-    signupData
-      .map(SignupData::isValid)
-      .compose(bindToLifecycle())
-      .subscribe(this.formIsValid);
-
-    signupData
-      .compose(takeWhen(this.signupClick))
-      .flatMap(this::submit)
-      .compose(bindToLifecycle())
-      .subscribe(this::success);
-
-    this.currentConfig.observable()
-      .take(1)
-      .map(config -> I18nUtils.isCountryUS(config.countryCode()))
-      .compose(bindToLifecycle())
-      .subscribe(this.sendNewslettersIsChecked::onNext);
-
-    this.signupError
-      .compose(bindToLifecycle())
-      .subscribe(__ -> this.koala.trackRegisterError());
-
-    this.sendNewslettersClick
-      .compose(bindToLifecycle())
-      .subscribe(this.koala::trackSignupNewsletterToggle);
-
-    this.signupSuccess
-      .compose(bindToLifecycle())
-      .subscribe(__ -> {
-        this.koala.trackLoginSuccess();
-        this.koala.trackRegisterSuccess();
-      });
-
-    this.koala.trackRegisterFormView();
+    /** Call when the signup button has been clicked. */
+    void signupClick();
   }
 
-  private Observable<AccessTokenEnvelope> submit(final @NonNull SignupData data) {
-    return this.client.signup(data.fullName, data.email, data.password, data.password, data.sendNewsletters)
-      .compose(Transformers.pipeApiErrorsTo(this.signupError))
-      .compose(Transformers.neverError())
-      .doOnSubscribe(() -> this.formSubmitting.onNext(true))
-      .doAfterTerminate(() -> this.formSubmitting.onNext(false));
+  interface Outputs {
+    /** Emits a string to display when signup fails. */
+    Observable<String> errorString();
+
+    /** Emits a boolean that determines if the sign up button is enabled. */
+    Observable<Boolean> formIsValid();
+
+    /** Emits a boolean that determines if the sign up button is disabled. */
```

Comment:
Does this comment line up with the name of the output? (sign up button disabled vs form submitting)

Reply:
I had a question about output javadoc. Should they be what's happening in the Activity or what the ViewModel is doing?

Reply:
chatted irl, and for posterity: in general an output should be doing the same thing in the ViewModel and the Activity so the javadoc would be representative of both

in this situation since we use both `formIsValid` and `formSubmitting` to set the enabled / disabled state of the signup button, [with funny lil helper methods](https://github.com/kickstarter/android-oss/blob/master/app/src/main/java/com/kickstarter/ui/activities/SignupActivity.java#L114-L120), we should ideally refactor to merge these outputs into one, e.g.

```java
Observable<Boolean> signupButtonIsEnabled();
```

and change the logic a bit in the VM, but this is a non-blocking change for purposes of this PR